src/structs.py

```diff
@@ -222,10 +222,16 @@ def variables(self):
         return list(self.entities)
 
     @cached_property
-    def _str(self):
+    def _str(self) -> str:
         return (str(self.predicate) + "(" +
                 ", ".join(map(str, self.variables)) + ")")
 
+    def ground(self, sub) -> GroundAtom:
```

Comment:
Please annotate the type of "sub" parameter

Reply:
so polite

Reply:
so polite